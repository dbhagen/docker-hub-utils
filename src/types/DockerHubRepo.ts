/**
 * This is a direct representation of what we get back from the `/repositories`
 * API call.
 */
export interface DockerHubAPIRepo {
  readonly can_edit: boolean
  readonly description: string
  readonly is_automated: boolean
  readonly is_migrated: boolean
  readonly is_private: boolean
  readonly last_updated: string
  readonly name: string
  readonly namespace: string
  readonly pull_count: number
  readonly repository_type: string
  readonly star_count: number
  readonly status: number
  readonly user: string
}

/**
 * Union type representing the architecture defined in part of an OCI image's
 * manifest list.
 *
 * As specified in the Docker Manifest spec, any valid GOARCH values are valid
 * image architecture values, and vice versa:
 * > The platform object describes the platform which the image in the manifest
 * > runs on. A full list of valid operating system and architecture values are
 * > listed in the Go language documentation for $GOOS and $GOARCH
 * @see https://docs.docker.com/registry/spec/manifest-v2-2/#manifest-list-field-descriptions
 */
export type Architecture =
  | '386'
  | 'amd64'
  | 'arm'
  | 'arm64'
  | 'mips'
  | 'mips64'
  | 'mips64le'
  | 'mipsle'
  | 'ppc64'
  | 'ppc64le'
  | 's390x'
  | 'wasm'

/**
 * Union type representing the OS defined in part of an OCI image's
 * manifest list.
 * See the docs for the `Architecture` type above for more info.
 */
export type OS =
  | 'aix'
  | 'android'
  | 'darwin'
  | 'dragonfly'
  | 'freebsd'
  | 'illumos'
  | 'js'
  | 'linux'
  | 'netbsd'
  | 'openbsd'
  | 'plan9'
  | 'solaris'
  | 'windows'

export enum ManifestMediaType {
  Manifest = 'application/vnd.docker.distribution.manifest.v2+json',
  ManifestList = 'application/vnd.docker.distribution.manifest.list.v2+json',
}

/**
 * Yes, there's *way* more information contained in the manifest / "fat"
 * manifestList than just architectures, but I find this to be the most
 * relevant section for my projects. PR's welcome.
 */
export interface DockerManifest {
  readonly digest: string
  readonly mediaType: ManifestMediaType
  readonly platform: Array<{
    os: OS
    architecture: Architecture
  }>
  readonly schemaVersion: 1 | 2 | any
}

export interface DockerManifestList {
  readonly manifests: DockerManifest[]
  readonly mediaType: ManifestMediaType
  readonly schemaVersion: 1 | 2 | any
}

export interface DockerHubRepo {
  // ========================
  // Main fields of interest
  // ========================
  readonly description: string | null | undefined
  // ISO8601-format string.
  readonly lastUpdated: string
  readonly name: string
  readonly pullCount: number
  readonly starCount: number
  readonly user: string

  // Manifest type *may* be nested within this interface, but is usually
  // fetched and returned separately.
  readonly manifestList?: DockerManifestList

  // =============================================
  // Other stuff that comes down through the API,
  // that some may find useful
  // =============================================
  readonly canEdit?: boolean
  readonly isAutomated?: boolean
  readonly isMigrated?: boolean
  readonly isPrivate?: boolean
  readonly namespace?: string
  readonly repositoryType?: string
  readonly status?: number
}
