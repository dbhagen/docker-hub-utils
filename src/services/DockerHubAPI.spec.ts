import axios from 'axios'
import R from 'ramda'

import manifestFixture from '../__tests__/manifest_fixture.json'
import fixture from '../__tests__/repos_fixture.json'
import { DockerHubRepo } from '../types/DockerHubRepo'
import {
  extractRepositoryDetails,
  fetchManifestList,
  queryTopRepos,
} from './DockerHubAPI'

const get = jest.spyOn(axios, 'get').mockResolvedValue({})

const repoFixtures = fixture.data.results

describe('DockerHub handler', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  test('extractRepositoryDetails for a non-existant repo', () => {
    const repositoryDetails = extractRepositoryDetails([])
    expect(repositoryDetails).toHaveLength(0)
  })

  test('extractRepositoryDetails happy path', () => {
    const repoResults = repoFixtures
    const repositoryDetails = extractRepositoryDetails(repoResults)
    expect(repositoryDetails).toHaveLength(R.length(repoResults))
    const keys = [
      'description',
      'lastUpdated',
      'name',
      'pullCount',
      'starCount',
    ]

    // Verify values are defined for each of the above keys.
    repositoryDetails.forEach(repo => {
      keys.forEach((key: string) => {
        expect(repo).toHaveProperty(key)
      })
    })
  })

  test('queryTopRepos', async () => {
    // Configure axios to return our fixtures, rather than hitting the API.
    get.mockResolvedValueOnce(fixture)

    const topRepos: DockerHubRepo[] = await queryTopRepos('jessestuart')

    // Only one HTTP call required -- no auth needed when querying repos.
    expect(get).toHaveBeenCalledTimes(1)
    expect(topRepos).toHaveLength(R.length(repoFixtures))

    const totalPulls = R.sum(R.pluck('pullCount', topRepos))
    expect(totalPulls).toMatchSnapshot()

    const topRepoName = R.pipe(
      R.head,
      R.path(['name']),
    )(topRepos)

    expect(topRepoName).toBe('owntracks')
  })

  test('fetchManifestList happy path.', async () => {
    get.mockResolvedValueOnce({ data: { token: 'FAKE_TOKEN' } })
    get.mockResolvedValueOnce(manifestFixture)

    const topRepo: DockerHubRepo = R.head(
      extractRepositoryDetails(repoFixtures),
    )
    if (!topRepo) {
      fail('Unable to load repos fixture.')
      return
    }

    const manifestList = await fetchManifestList(topRepo)
    expect(get).toHaveBeenCalledTimes(2)
    expect(get.mock.calls).toMatchSnapshot()
    expect(get).toHaveBeenNthCalledWith(1, 'https://auth.docker.io/', {
      params: {
        scope: 'repository:jessestuart/owntracks:pull',
        service: 'registry.docker.io',
      },
    })
    expect(get).toHaveBeenNthCalledWith(
      2,
      'https://registry-1.docker.io/v2/jessestuart/owntracks/manifests/latest',
      {
        headers: {
          Accept: 'application/vnd.docker.distribution.manifest.list.v2+json',
          Authorization: 'Bearer FAKE_TOKEN',
        },
      },
    )

    expect(manifestList).toMatchSnapshot()
  })

  test('fetchManifestList fails when no token returned.', async () => {
    get.mockResolvedValueOnce({ data: {} })
    get.mockResolvedValueOnce(manifestFixture)

    const topRepo = R.head(extractRepositoryDetails(repoFixtures))
    if (!topRepo) {
      fail('Unable to load repos fixture.')
      return
    }

    expect(fetchManifestList(topRepo)).rejects.toThrow()
    expect(get).toHaveBeenCalledTimes(1)
  })
})
