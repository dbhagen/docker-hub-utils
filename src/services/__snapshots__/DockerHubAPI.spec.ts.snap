// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DockerHub handler fetchManifestList happy path. 1`] = `
Array [
  Array [
    "https://auth.docker.io/",
    Object {
      "params": Object {
        "scope": "repository:jessestuart/owntracks:pull",
        "service": "registry.docker.io",
      },
    },
  ],
  Array [
    "https://registry-1.docker.io/v2/jessestuart/owntracks/manifests/latest",
    Object {
      "headers": Object {
        "Accept": "application/vnd.docker.distribution.manifest.list.v2+json",
        "Authorization": "Bearer FAKE_TOKEN",
      },
    },
  ],
]
`;

exports[`DockerHub handler fetchManifestList happy path. 2`] = `
Object {
  "manifests": Array [
    Object {
      "digest": "sha256:d0593f2bf35b854611dda94b5210243f8473847150d1709b12163453841bf5f8",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": Object {
        "architecture": "amd64",
        "os": "linux",
      },
      "size": 1365,
    },
    Object {
      "digest": "sha256:0ec4c38328ab0e0b8798c28384fd44bab385b8313b13aa8f494e940e83a12d69",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": Object {
        "architecture": "arm",
        "os": "linux",
      },
      "size": 1369,
    },
    Object {
      "digest": "sha256:72d730063d55fa164378188f92119e00769dbc759f95cbf1319705e90d6214ea",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": Object {
        "architecture": "arm64",
        "os": "linux",
      },
      "size": 1369,
    },
  ],
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
  "schemaVersion": 2,
}
`;

exports[`DockerHub handler queryTopRepos 1`] = `1718496`;
